drivers/FlashIAP.cpp

```diff
@@ -126,9 +143,14 @@ int FlashIAP::program(const void *buffer, uint32_t addr, uint32_t size)
             prog_buf = buf;
             prog_size = chunk;
         }
-        if (flash_program_page(&_flash, addr, prog_buf, prog_size)) {
-            ret = -1;
-            break;
+        {
+            CriticalSectionLock critical_section;
```

Comment:
Are we sure about this critical section? I fear this could break some platforms.

We've had issues in the past with systems that would disable interrupts because they couldn't run from the same bank they were flashing storage in.  And in the end those systems just didn't work - the interrupt disabling killed the radio driver. They were forced to ensure enough bank separation to allow us keep the radio driver IRQs running while programming the other bank. (Eg mbed client on K64F must keep its image in the first half, flash storage in the second half, if using program flash for storage).

I'm not sure what the current situation is - whether there are systems using FlashIAP for a storage area that expect no interrupt disruption during access.

Looking at some `flash_program_page` implementations, I see them already calling `core_util_critical_section_enter` themselves. If HALs can call that, should they maybe be expected to call `mbed_mpu_manager_lock_rom_write` and `mbed_mpu_manager_lock_ram_execution` themselves, if necessary?

Anyone who needs RAM execution must also already be enabling a critical section, so they could do both themselves. Temporarily allowing just the ROM writing for people not using a RAM programmer and keeping IRQs on is not the end of the world. Then we wouldn't be changing the IRQ disabling profile for any platform.

At the minute the API does permit people to leave RAM execution or ROM writing on while others are running. If you wanted to rule either out totally, maybe a critical section claim should be built-in to the lock call(s).

Reply:
I'll remove the critical section from this PR as it is not a requirement for MPU support. If need be I can create it in a separate PR where we can have a more focused discussion.